dconf2nix: init at 0.5.0

gvolpe opened the pull request:
###### Motivation for this change

To make the [dconf2nix](https://github.com/gvolpe/dconf2nix) package available on Nixpkgs.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

gvolpe committed:
dconf2nix: init at 0.5.0

gvolpe commented:
Apparent'y I'm not a trusted user: https://gist.github.com/GrahamcOfBorg/755c9a6d20cccee353811b1df5fb430a 

```
nix-env failed:
warning: ignoring the user-specified setting 'restrict-eval', because it is a restricted setting and you are not a trusted user
```

If there's anything I should fix, I'd appreciate any help :pray:

jonringer commented on maintainers/maintainer-list.nix:
```diff
@@ -3218,6 +3218,12 @@
     githubId = 6768842;
     name = "Joris Guyonvarch";
   };
+  gvolpe = {
+    email = "volpegabriel@gmail.com";
+    github = "gvolpe";
+    githubId = 443978;
+    name = "Gabriel Volpe";
+  };
```
Please make this into a separate commit:

```
maintainers: add <name>
```

jonringer commented on pkgs/development/tools/haskell/dconf2nix/default.nix:
```diff
@@ -0,0 +1,21 @@
+}:
+mkDerivation rec {
+  pname   = "dconf2nix";
+  version = "0.0.5";
```
linting, sorry for the nits
```suggestion
  version = "0.0.5";

```

jonringer commented on pkgs/development/tools/haskell/dconf2nix/default.nix:
```diff
@@ -0,0 +1,21 @@
+{ mkDerivation, base, containers, optparse-applicative, parsec
+, stdenv, text
+}:
```
```suggestion
}:

```

jonringer commented on pkgs/development/tools/haskell/dconf2nix/default.nix:
```diff
@@ -0,0 +1,21 @@
+  src = builtins.fetchTarball {
+    url    = "https://github.com/gvolpe/dconf2nix/archive/v${version}.tar.gz";
+    sha256 = "0immbx4bgfq3xmbbrpw441nx0sdpm4cp64s7qbvcbvllp4gbivpg";
+  };
```
```suggestion
  };

```

jonringer commented on pkgs/development/tools/haskell/dconf2nix/default.nix:
```diff
@@ -0,0 +1,21 @@
+  libraryHaskellDepends = [
+    base containers optparse-applicative parsec text
+  ];
+  executableHaskellDepends = [ base ];
```
```suggestion
  executableHaskellDepends = [ base ];

```

jonringer commented on pkgs/development/tools/haskell/dconf2nix/default.nix:
```diff
@@ -0,0 +1,21 @@
+  executableHaskellDepends = [ base ];
+  homepage = "https://github.com/gvolpe/dconf2nix";
+  description = "Convert dconf files to Nix, as expected by Home Manager";
+  license     = stdenv.lib.licenses.asl20;
+  maintainers = with stdenv.lib.maintainers; [ gvolpe ];
```
personal preference, since lib is so common
```suggestion
  license     = lib.licenses.asl20;
  maintainers = with lib.maintainers; [ gvolpe ];
```

jonringer commented on pkgs/development/tools/haskell/dconf2nix/default.nix:
```diff
@@ -0,0 +1,21 @@
+{ mkDerivation, base, containers, optparse-applicative, parsec
+, stdenv, text
```
```suggestion
, lib, text
```

jonringer commented:
> Apparently, I'm not a trusted user: https://gist.github.com/GrahamcOfBorg/755c9a6d20cccee353811b1df5fb430a
> 
> ```
> nix-env failed:
> warning: ignoring the user-specified setting 'restrict-eval', because it is a restricted setting and you are not a trusted user
> ```
> 
> If there's anything I should fix, I'd appreciate any help pray

This is because you're doing a IFD (import from derivation)

jonringer commented on pkgs/development/tools/haskell/dconf2nix/default.nix:
```diff
@@ -0,0 +1,21 @@
+mkDerivation rec {
+  pname   = "dconf2nix";
+  version = "0.0.5";
+  src = builtins.fetchTarball {
```
```suggestion
  src = fetchurl {
```

gvolpe commented:
Thanks for the quick review @jonringer ! I'll address your comments soon.

> This is because you're doing a IFD (import from derivation)

Does this have any implications? Should I do it differently?

jonringer reviewed (commented):
The main issue is that you're trying to evaluate ANOTHER derivation from the src fixed output derivation.

Please move all the relevant packaging detail from upstream to here.

@cdepillabout please help with haskell packaging :)

gvolpe commented:
I think I addressed all your comments, @jonringer , thanks again!

> The main issue is that you're trying to evaluate ANOTHER derivation from the src fixed output derivation.

Not sure exactly how I should solve this, any help would be much appreciated :pray:

gvolpe commented:
Maybe doing somethink like this?

```nix
prePatch = ''
  rm default.nix 
'';
```

cdepillabout commented on pkgs/development/haskell-modules/non-hackage-packages.nix:
```diff
@@ -10,6 +10,8 @@ self: super: {
   multi-ghc-travis = throw ("haskellPackages.multi-ghc-travis has been renamed"
     + " to haskell-ci, which is now on hackage");
 
+  dconf2nix = self.callPackage ../tools/haskell/dconf2nix { };
```
Rather than trying to build this package as a `non-hackage-package`, it would be easier for us if you were to just release it to Hackage.

If `dconf2nix` was released to Hackage, then the Haskell tooling here in Nixpkgs will pull in the latest version everytime you make a new release to Hackage.

Also, releasing to Hackage generally helps the wider Hackage community, since non-nixpkg users have an easy way to get your package.

gvolpe replied:
I know about the trade-offs but if we think about it, this is a tool for Nix users so I'd expect anyone using it to install it via Nix and not by other means.

OTOH I don't want to maintain a Hackage package, that's why I added myself as a maintainer here.

cdepillabout replied:
If you don't upload to Hackage, it will probably be more work for you to maintain it here in nixpkgs, but as long as you're willing to deal with that extra work, it is fine.

Check out how `haskellPackages.spago` is packaged in nixpkgs, and model the packaging for `dconf2nix` on that.

gvolpe replied:
I might consider it in the long run if this doesn't work out but for now I just want to make it available on Nixpkgs as a top level package, so the PR must exist regardless of being on Hackage or not.

I looked at the definition of [vaultenv](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/haskell/vaultenv/default.nix) to have an idea about how to package it but it's also my first time doing it so I'm not really sure what else I'm missing. I'll have a look at Spago too :+1:

gvolpe commented:
I don't really understand how Spago is built :confused: . I see it has a [similar derivation](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/tools/purescript/spago/spago.nix) but on the Github repository there's no `default.nix` file.

OTOH `vaultenv` has a very simple [default.nix](https://github.com/channable/vaultenv/blob/master/default.nix) in the Github repository but it uses Static Haskell Nix which is only supported by Stack AFAIU and I'm using Cabal.

I think the issue here with my derivation is due to my [default.nix](https://github.com/gvolpe/dconf2nix/blob/master/default.nix) calling `cabal2nix`? If so, any suggestions on what's the approach I should follow? Because I need that `default.nix` to build the project locally and on the CI build, and the `shell.nix` for my local dev env.

Any help much appreciated :bow:

cdepillabout commented:
@gvolpe You can basically copy the `spago` directory here in nixpkgs and re-use it for `dconf2nix`.

Basically, you'll need a `dconf2nix.nix` file that is generated by manually calling `cabal2nix`.  You'll probably also want a `default.nix` file that adds some `meta` values to `haskellPackages.dconf2nix`, as well as doing things like generating optparse-applicative autocomplete stuff.  Finally, you'll need an `update.sh` script that will make it easy to update the `dconf2nix.nix` file.

> I see [spago] has a similar derivation but on the Github repository there's no default.nix file.

Yeah, there are derivations for building Haskell packages here in nixpkgs, but there are no derivations on most of the upstream Haskell packages.  In general you will want to look for how to package stuff here in nixpkgs, not in the upstream repos.

> I think the issue here with my derivation is due to my default.nix calling cabal2nix? If so, any suggestions on what's the approach I should follow? Because I need that default.nix to build the project locally and on the CI build, and the shell.nix for my local dev env.

Yes, the problem is that you're using IFD: https://blog.hercules-ci.com/2019/08/30/native-support-for-import-for-derivation/

The approach that you should follow is the exact same approach that `spago` is using here in nixpkgs.

Because Hydra doesn't allow IFD, you will need to duplicate the Nix files here in nixpkgs and upstream.  This is somewhat unfortunate, but it is the current state of affairs here in nixpkgs.

gvolpe commented:
Thanks @cdepillabout , I'll have some time tomorrow to look into it.

> In general you will want to look for how to package stuff here in nixpkgs, not in the upstream repos.

FWIW I'm not packing it per-se but I'm taking advantages of Cachix and Github Actions. Isn't there something like `fetchTarball` but that can ignore the `default.nix` files so it avoids IFD? That seems like a rather common scenario. 

Also, the current `default.nix` in this PR builds fine locally by running `nix-build -A dconf2nix`. Is there a way to verify there is no IFD happening? So I can test it locally.

Thanks again for your time!

gvolpe commented:
I tried to follow how `spago` is built but now I'm no longer sure how to build it locally. `nix-build -A dconf2nix` no longer works due to having some dependencies in the new `default.nix` (more or less copied from `spago`s):

```nix
{ haskell, haskellPackages, lib, runCommand }:

let
  dconf2nix =
    haskell.lib.justStaticExecutables
      (haskell.lib.overrideCabal haskellPackages.dconf2nix (oldAttrs: {
        maintainers = (oldAttrs.maintainers or []) ++ [
          lib.maintainers.cdepillabout
        ];
      }));
in

dconf2nix.overrideAttrs (oldAttrs: {
  passthru = (oldAttrs.passthru or {}) // {
    updateScript = ./update.sh;

    # These tests can be run with the following command.  The tests access the
    # network, so they cannot be run in the nix sandbox.  sudo is needed in
    # order to change the sandbox option.
    #
    # $ sudo nix-build -A dconf2nix.passthru.tests --option sandbox relaxed
    #
    tests =
      runCommand
        "dconf2nix-tests"
        {
          __noChroot = true;
          nativeBuildInputs = [
            dconf2nix
          ];
        }
        ''
          dconf2nix --v
        '';
  };
})
```

jtojnar commented on pkgs/development/tools/haskell/dconf2nix/default.nix:
```diff
@@ -0,0 +1,25 @@
+  pname   = "dconf2nix";
+  version = "0.0.5";
+
+  src = builtins.fetchurl {
```
GitHub-generated archives can change hashes. Use `fetchFromGitHub`, which hashes the archive’s content.

cdepillabout commented:
> I tried to follow how spago is built but now I'm no longer sure how to build it locally. nix-build -A dconf2nix no longer works due to having some dependencies in the new default.nix (more or less copied from spagos)

Would you be able to push what you have so far?  The code snippet you pasted looks fine for the most part, so I am having trouble seeing what you're having trouble with.

gvolpe commented:
Sure @cdepillabout , I just pushed what I have, based on the `spago` derivation. I don't know how to test it, though.

gvolpe commented:
The command `sudo nix-build -A spago.passthru.tests --option sandbox relaxed` (located in the comments of `update.sh`) fails for me with the same error I mentioned above due to `default.nix` having some dependencies.

```sh
error: cannot auto-call a function that has an argument without a default value ('haskell')
```

cdepillabout commented:
@gvolpe Here's what I needed to do to get this to build:

```diff
iff --git a/pkgs/development/haskell-modules/non-hackage-packages.nix b/pkgs/development/haskell-modules/non-hackage-packages.nix
index dd6a13f23c2..8801f1f1ddd 100644
--- a/pkgs/development/haskell-modules/non-hackage-packages.nix
+++ b/pkgs/development/haskell-modules/non-hackage-packages.nix
@@ -10,7 +10,7 @@ self: super: {
   multi-ghc-travis = throw ("haskellPackages.multi-ghc-travis has been renamed"
     + " to haskell-ci, which is now on hackage");
 
-  dconf2nix = self.callPackage ../tools/haskell/dconf2nix { };
+  dconf2nix = self.callPackage ../tools/haskell/dconf2nix/dconf2nix.nix { };
 
   # https://github.com/channable/vaultenv/issues/1
   vaultenv = self.callPackage ../tools/haskell/vaultenv { };
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index f3ec07d8534..dcff5fadd66 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -2921,7 +2921,7 @@ in
 
   dclxvi = callPackage ../development/libraries/dclxvi { };
 
-  dconf2nix = haskellPackages.dconf2nix;
+  dconf2nix = callPackage ../development/tools/haskell/dconf2nix { };
 
   dcraw = callPackage ../tools/graphics/dcraw { };
```

gvolpe commented:
Wow, I feel dumb I didn't see that! Thanks @cdepillabout, I'll try it out as soon as I get home 🙇

gvolpe commented:
@cdepillabout what command do you run to build `dconf2nix`? I made the changes you suggested but I'm still not able to build it by running `nix-build -A dconf2nix` or `sudo nix-build -A dconf2nix.passthru.tests --option sandbox relaxed`.

gvolpe commented:
Oh nevermind, it builds fine. I was running the command not from the top-level directory... Thanks a lot for your help!

cdepillabout reviewed (changes requested):
There are a few other small fixes that can be made:

```diff
diff --git a/pkgs/development/tools/haskell/dconf2nix/default.nix b/pkgs/development/tools/haskell/dconf2nix/default.nix
index 3bcd4e2e1f3..cbaa47b6556 100644
--- a/pkgs/development/tools/haskell/dconf2nix/default.nix
+++ b/pkgs/development/tools/haskell/dconf2nix/default.nix
@@ -5,7 +5,7 @@ let
     haskell.lib.justStaticExecutables
       (haskell.lib.overrideCabal haskellPackages.dconf2nix (oldAttrs: {
         maintainers = (oldAttrs.maintainers or []) ++ [
-          lib.maintainers.cdepillabout
+          lib.maintainers.gvolpe
         ];
       }));
 in
@@ -14,23 +14,19 @@ dconf2nix.overrideAttrs (oldAttrs: {
   passthru = (oldAttrs.passthru or {}) // {
     updateScript = ./update.sh;
 
-    # These tests can be run with the following command.  The tests access the
-    # network, so they cannot be run in the nix sandbox.  sudo is needed in
-    # order to change the sandbox option.
-    #
-    # $ sudo nix-build -A dconf2nix.passthru.tests --option sandbox relaxed
+    # These tests can be run with the following command.
     #
+    # $ nix-build -A dconf2nix.passthru.tests
     tests =
       runCommand
         "dconf2nix-tests"
         {
-          __noChroot = true;
           nativeBuildInputs = [
             dconf2nix
           ];
         }
         ''
-          dconf2nix -v
+          dconf2nix > $out
         '';
   };
 })
```

You don't need to run the tests with the relaxed sandbox, because it appears that just running `dconf2nix` produces output.  I'm assuming this is a good enough check to see if `dconf2nix` is compiled correctly, but if there is a better one please feel free to use it.

I've made sure that `dconf2nix` builds, the `update.sh` script works as expected, and the tests pass.  Once you make these above changes, this looks good and I will go ahead and merge it in!

_edit_: Oh, and usually other maintainers appreciate if you squash your commits, and make sure that the commit adding you as a maintainer comes before the commit where you add `dconf2nix`.

gvolpe committed:
maintainers: add gvolpe

cdepillabout reviewed (approved):
Looks good, thanks for working through this with us!

gvolpe commented:
Thank you for being patient @cdepillabout ! :bow:
